File: app/Libraries/Morphology/ContextRules.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.IO;
using System.Diagnostics;

namespace MorfParsingLibrary
{
    /// <summary>
    /// Makes something important
    /// </summary>
    class ContextRules
    {
        //C-согласная
        //F-мягкая гласная
        //B-твердая гласная
        public static string B = "аӑуы";
        public static string C = "бвгджзйклмнпрсҫтфхцчшщ";
        public static string F = "еӗÿя";
        public static string WORD;
        public static string AFFUNIT;


        //поиск правила обработки контекстов
        public static void FindTemplate(string word, string AffUnit, string[] rules,ref string[] syllables, ref string chastrechi)
        {
            WORD = word;
            AFFUNIT = AffUnit;

            string lcon = word.Length > 1 ? word.Substring(word.Length - 2) : word; //получение лев контекста            
            string rcon = AffUnit.Split(new char[] { '|' }, StringSplitOptions.RemoveEmptyEntries)[0]; //получение пр контекста

            foreach (string line in rules)
            {
                if (line.StartsWith("//")) continue; // пропуск комментариев
                string[] rule = line.Split(new char[] { ';' }); //читаем строку с правилом
                if (CheckLeftContext(lcon, rule[0]) && CheckRightContext(rcon, rule[1]))//если найдено подходящ правило
                {
                    //согласно правилу
                    WordModify(word, rule[2]); //слово
                    
                    syllables = ExtractSyllables(rule[3]);//символы восстановления
                    chastrechi = rule[4];//

                    Trace.WriteLine("Правило контекста: [" + WORD + " " + AFFUNIT + " | " + line + "]");
                    return;
                }

            }
            Trace.WriteLine("No Context Rule");

        }
        //символы по умолчанию
        public static string[] DefaultSyllables(string AffUnit)
        {
            string[] syll = null;
            switch (AffixInfoProvider.TypeOfRecovery(AffUnit))
            {
                case AffixInfoProvider.RecoveryMode.With:
                    syll = new string[] { "а", "е", "ӗ", "я" };
                    break;
                case AffixInfoProvider.RecoveryMode.Without:
                    syll = new string[] { "" };
                    break;
                case AffixInfoProvider.RecoveryMode.Both:
                    syll = new string[] { "а", "е", "ӗ", "я",  "" };//ӑ
                    break;
                    
            }
            return syll;
        }
        //проверка подходимости части речи
        public static bool CheckChastRechi(string chastrechi, string context_chastrechi)
        {
            if (context_chastrechi == null) return false;
            bool b = false;
            switch (context_chastrechi)
            {
                case "":
                    b = true;
                    break;
                default:
                    string[] m = context_chastrechi.Split(new char[] { ',' });
                    for (int i = 0; i < m.Length; i++)
                    {
                        string turned = Helper.convertPartOfSpeech(m[i]);
                        if (turned == chastrechi) 
                        {
                            b = true; 
                            break;
                        } 
                    }
                    break;
            }
            return b;
        }
        //формирование массива символов
        private static string[] ExtractSyllables(string p)
        {
            if (p == "") return new string[] { "" };
            string[] res = p.Split(new char[] { ',' }, StringSplitOptions.RemoveEmptyEntries);
            return res;
        }
        //редактирование аффикса 
        private static void AffixModify(string AffUnit, string p)
        {
            throw new NotImplementedException();
        }
        //редактирование слова согласно правилу
        private static void WordModify(string word, string p)
        {
            if (p == "") return;
            switch (p[0])
            {
                case '-':
                    WORD = word.Substring(0, word.Length - (int)Char.GetNumericValue(p[1]));
                    break;
                default:
                    break;
            }
            
        }
        //проверка лев контекста
        private static bool CheckLeftContext(string lcon_input, string lcon_rule)
        {
            bool compatible = true;
            if (lcon_rule != "")
            {
                if ((lcon_input != lcon_rule))
                {
                    if (!CompatibleLeft(lcon_input, lcon_rule))
                    {
                        compatible = false;
                    }
                }
            }
            return compatible;
        }
        //проверка пр контекста
        private static bool CheckRightContext(string rcon_input, string rcon_rule)
        {
            bool compatible = true;
            if (rcon_rule != "")
            {
                if (rcon_input != rcon_rule)
                {
                    if (!CompatibleRight(rcon_input, rcon_rule))
                    {
                        compatible = false;
                    }
                }
            }


            return compatible;
        }
        //проверка совместимости лев
        private static bool CompatibleLeft(string input, string rule)
        {

            string[] mas = rule.Split(new char []{','}, StringSplitOptions.RemoveEmptyEntries);
            for (int i = 0; i < mas.Length; i++)
            {
                if (input == mas[i])
                {
                    return true;
                }
            }


            string context = "";
            if (input.Length > 1)
            {
                if (C.Contains(input[0]))
                {
                    if (input[0] == input[1])
                    {
                        context = "2C";
                    }
                    else
                    {
                        context = "C" + input[1];
                    }
                } 
            }
            
            bool b = context == rule;
            return b;
        }
        //проверка совместимости прав
        private static bool CompatibleRight(string input, string rule)
        {
            string[] mas = rule.Split(new char[] { ',' }, StringSplitOptions.RemoveEmptyEntries);
            for (int i = 0; i < mas.Length; i++)
            {
                if (input == mas[i])
                {
                    return true;
                }
            }

            string context = "";
            if (F.Contains(input[0]))
            {
                context = "F";
            }
            else if (B.Contains(input[0]))
            {
                context = "B";
            } 


            bool b = context == rule;
            return b;
        }




        //проверка тверд/мягкость слова& 
        //true - тверд
        internal static bool Consistency(string word)
        {
            char x = word[word.Length - 1];

            string Back = "аӑуыо";
            string Front = "еӗиÿэюя";
            string special = "ьъ";
            
            if (special.Contains(x)) return false;

            for (int i = word.Length - 1; i >= 0; i--)
            {
                if (Back.Contains(word[i]))
                {
                    //Trace.WriteLine(word[i]);
                    return true; //твердыня
                }
                    
                if (Front.Contains(word[i]))
                {
                    //Trace.WriteLine(word[i]);
                    return false; //мягкота
                }
                    
            }
            throw new NotImplementedException("WTF man?");
        }
        //оканчив. на согл. или нет
        internal static bool Soglasnaya(string word)
        {
            bool sogl = false;
            string S = "бвгджзйклмнпрсҫтфхцчшщ";
            string special = "ьъ";
            char x = word[word.Length - 1];
            
            sogl = (S.Contains(x) || (special.Contains(x))) ? true : false; //оканчивается ли слово на согл
            return sogl;
        }
        //проверка звонкости согласной буквы
        internal static bool Zvonko(char character)
        {
            string z = "гджзйлмнр";
            //string nez = "бвкптфхцчшщ";

            if (z.Contains(character)) return true;
            else return false;
            throw new NotImplementedException("WTF man?");
        }
    }
}

File: app/Services/MorfParsingLibrary/ContextRules.php
<?php
namespace App\Services\MorfParsingLibrary;


class ContextRules {
    // C-согласная
    // F-мягкая гласная
    // B-твердая гласная
    public static $B = "аӑуы";
    public static $C = "бвгджзйклмнпрсҫтфхцчшщ";
    public static $F = "еӗÿя";
    public static $WORD;
    public static $AFFUNIT;

    // поиск правила обработки контекстов
    public static function FindTemplate($word, $AffUnit, $rules, &$syllables, &$chastrechi) {
        self::$WORD = $word;
        self::$AFFUNIT = $AffUnit;

        $lcon = mb_strlen($word) > 1 ? mb_substr($word, mb_strlen($word) - 2) : $word; // получение лев контекста
        $affunit_parts = explode('|', $AffUnit);
        $rcon = $affunit_parts[0]; // получение пр контекста

        foreach ($rules as $line) {
            if (strpos($line, "//") === 0) continue; // пропуск комментариев
            $rule = explode(';', $line); // читаем строку с правилом
            if (self::CheckLeftContext($lcon, $rule[0]) && self::CheckRightContext($rcon, $rule[1])) {
                // согласно правилу
                self::WordModify($word, $rule[2]); // слово
                $syllables = self::ExtractSyllables($rule[3]); // символы восстановления
                $chastrechi = $rule[4]; //

                echo "Правило контекста: [" . self::$WORD . " " . self::$AFFUNIT . " | " . $line . "]<br>";
                return;
            }
        }
        echo "No Context Rule<br>";
    }

    // символы по умолчанию
    public static function DefaultSyllables($AffUnit) {
        switch (AffixInfoProvider::TypeOfRecovery($AffUnit)) {
            case AffixInfoProvider::RecoveryMode['With']:  // Обращение к элементу массива-константы
                return ["а", "е", "ӗ", "я"];
            case AffixInfoProvider::RecoveryMode['Without']:
                return [""];
            case AffixInfoProvider::RecoveryMode['Both']:
                return ["а", "е", "ӗ", "я", ""];
        }
        return [];
    }

    // проверка подходимости части речи
    public static function CheckChastRechi($chastrechi, $context_chastrechi)
     {
        if ($context_chastrechi === null) return false;
        switch ($context_chastrechi) {
            case "":
                return true;
            default:
                $m = explode(',', $context_chastrechi);
                foreach ($m as $item) {
                    $turned = Helper::convertPartOfSpeech($item);
                    if ($turned == $chastrechi) {
                        return true;
                    }
                }
                break;
        }
        return false;
    }

    // формирование массива символов
    private static function ExtractSyllables($p) {
        if ($p == "") return [""];
        return explode(',', $p);
    }

    // редактирование аффикса
    private static function AffixModify($AffUnit, $p) {
        throw new \LogicException("Not implemented");
    }

    // редактирование слова согласно правилу
    private static function WordModify(&$word, $p) {
        if ($p == "") return;
        switch ($p[0]) {
            case '-':
                $length = (int)$p[1];
                $word = mb_substr($word, 0, mb_strlen($word) - $length);
                break;
            default:
                break;
        }
    }

    // проверка лев контекста
    private static function CheckLeftContext($lcon_input, $lcon_rule) {
        if ($lcon_rule == "") return true;
        if ($lcon_input != $lcon_rule) {
            if (!self::CompatibleLeft($lcon_input, $lcon_rule)) {
                return false;
            }
        }
        return true;
    }

    // проверка пр контекста
    private static function CheckRightContext($rcon_input, $rcon_rule) {
        if ($rcon_rule == "") return true;
        if ($rcon_input != $rcon_rule) {
            if (!self::CompatibleRight($rcon_input, $rcon_rule)) {
                return false;
            }
        }
        return true;
    }

    // проверка совместимости лев
    private static function CompatibleLeft($input, $rule) {
        $mas = explode(',', $rule);
        foreach ($mas as $item) {
            if ($input == $item) {
                return true;
            }
        }

        $context = "";
        if (mb_strlen($input) > 1) {
            if (strpos(self::$C, $input[0]) !== false) {
                if ($input[0] == $input[1]) {
                    $context = "2C";
                } else {
                    $context = "C" . $input[1];
                }
            }
        }

        return $context == $rule;
    }

    // проверка совместимости прав
    private static function CompatibleRight($input, $rule) {
        $mas = explode(',', $rule);
        foreach ($mas as $item) {
            if ($input == $item) {
                return true;
            }
        }

        $context = "";
        if (strpos(self::$F, $input[0]) !== false) {
            $context = "F";
        } elseif (strpos(self::$B, $input[0]) !== false) {
            $context = "B";
        }

        return $context == $rule;
    }

    // проверка тверд/мягкость слова& 
    // true - тверд
    public static function Consistency($word) {
        $x = $word[mb_strlen($word) - 1];

        $Back = "аӑуыо";
        $Front = "еӗиÿэюя";
        $special = "ьъ";

        if (strpos($special, $x) !== false) return false;

        for ($i = mb_strlen($word) - 1; $i >= 0; $i--) {
            if (strpos($Back, $word[$i]) !== false) {
                return true; // твердыня
            }

            if (strpos($Front, $word[$i]) !== false) {
                return false; // мягкота
            }
        }
     
        throw new \LogicException("WTF man?");
    }

    // оканчив. на согл. или нет
    public static function Soglasnaya($word) {
        $sogl = false;
        $S = "бвгджзйклмнпрсҫтфхцчшщ";
        $special = "ьъ";
        $x = $word[mb_strlen($word) - 1];

        $sogl = (strpos($S, $x) !== false || strpos($special, $x) !== false); // оканчивается ли слово на согл
        return $sogl;
    }

    // проверка звонкости согласной буквы
    public static function Zvonko($character) {
        $z = "гджзйлмнр";
        // $nez = "бвкптфхцчшщ";

        return strpos($z, $character) !== false;
    }
}